internal/server/resource.go

```diff
@@ -849,6 +854,47 @@ func resourceServerDelete(ctx context.Context, d *schema.ResourceData, m interfa
 		d.SetId("")
 		return nil
 	}
+
+	var warnings diag.Diagnostics
+
+	if d.Get("shutdown_before_deletion").(bool) {
+		// Try shutting down the server
+		shutdownResult, _, err := client.Server.Shutdown(ctx, &hcloud.Server{ID: serverID})
+		if err != nil {
+			return hcclient.ErrorToDiag(err)
+		}
+
+		if err = hcclient.WaitForAction(ctx, &client.Action, shutdownResult); err != nil {
+			return hcclient.ErrorToDiag(err)
+		}
+
+		// Give the server some time to shut down
+		err = control.Retry(control.DefaultRetries, func() error {
+			server, _, err := client.Server.GetByID(ctx, serverID)
+
+			// If it is not possible to get the server status, it is probably futile to retry
+			if err != nil {
+				return control.AbortRetry(err)
+			}
+
+			if server.Status != hcloud.ServerStatusOff {
+				return fmt.Errorf("Server has not shut down yet")
+			}
+
+			// Server has shut down successfully
+			return nil
+		})
+		if err != nil {
+			// If shutting down does not work, add a warning and move on with deletion
+			warnings = append(warnings, diag.Diagnostic{
+				Severity: diag.Warning,
+				Summary:  fmt.Sprintf("Server id %d took longer than 30s to shut down gracefully, deleting it anyways.", serverID),
+			})
```

Comment:
Nice!

Reply:
I actually re-read the text because of that comment and changed it, removing the time, as I am using the retry function's defaults instead of a set timer.